src/main/java/org/mozilla/iot/webthing/Property.java

```diff
@@ -63,9 +65,122 @@ public Property(Thing thing,
         this.value.addObserver((a, b) -> this.thing.propertyNotify(this));
     }
 
-    private boolean readOnly() {
-        return this.metadata.containsKey("readOnly") &&
-                (boolean)this.metadata.get("readOnly");
+    /**
+     * Validate new property value before setting it.
+     *
+     * @param value New value
+     * @throws PropertyError On validation error.
+     */
+    private void validateValue(T value) throws PropertyError {
+        if (this.metadata.containsKey("type")) {
+            switch ((String)this.metadata.get("type")) {
+                case "null":
+                    if (!JSONObject.NULL.equals(value)) {
+                        throw new PropertyError("Value must be null");
+                    }
+                    break;
+                case "boolean":
+                    if (!(Boolean.FALSE.equals(value) ||
+                                  Boolean.TRUE.equals(value) ||
+                                  (value instanceof String &&
+                                           (((String)value).equalsIgnoreCase(
+                                                   "true") ||
+                                                    ((String)value).equalsIgnoreCase(
+                                                            "false"))))) {
+                        throw new PropertyError("Value must be a boolean");
+                    }
+                    break;
+                case "object":
+                    if (!(value instanceof JSONObject)) {
+                        throw new PropertyError("Value must be an object");
+                    }
+                    break;
+                case "array":
+                    if (!(value instanceof JSONArray)) {
+                        throw new PropertyError("Value must be an array");
+                    }
+                    break;
+                case "number":
+                    if (!(value instanceof Number)) {
+                        throw new PropertyError("Value must be a number");
+                    }
+                    break;
+                case "integer":
+                    if (!(value instanceof Number)) {
+                        throw new PropertyError("Value must be an integer");
+                    }
+
+                    double v = ((Number)value).doubleValue();
+                    if (v != Math.round(v)) {
```

Comment:
I believe this needs a tolerance or epsilon value to deal with potential rounding issues. In our likely range of values this might not be a problem though.

Reply:
Good catch. Done.